### ISSUE TITLE
Security entitty, better approach?

### ISSUE THREAD

**Issue by apryiomka** [vhrobles/NET-Core--EF7-with-Angular#4]
You have one Security entity with multiple security types. Some security types can have unique properties that are not available to other security types. For instance, Bonds would have interest rates, where as Stocks would have dividends. Would you have better approach defining your entities and the DB schema (via EF)?

**Comment by vhrobles** [comment]
Sure. I could add a catalog table [SecurityTypeAttributes] with a relationship of 1 to many from SecurityType, where I could store different properties (such as the interest rates and dividends) and its value, remove Type from SecurityType and add it to [SecurityTypeAttributes] as foreign key.

**Comment by apryiomka** [comment]
I was actually looking for inheritance - base security class and derived classes such as Bonds, Stocks and Funds which in turn would be different tables in the DB:

> public abstract class Security : Entity<int>
>     {
>         public string Simbol { get; set; }
>         public virtual ICollection<Portfolio> Protfolios { get; set; } 
>         public abstract double GetSecurityValue();
>     }
>     public class Stock : Security
>     {
>         public override double GetSecurityValue()
>         {
>             return UnitPrice;
>         }
>         public string CompanyName { get; set; }
>         public double UnitPrice { get; set; }
>     }
>     public class Fund : Security
>     {
>         public override double GetSecurityValue()
>         {
>             return 999;
>         }
>         public string FundType { get; set; }
>     }
>     public class Bond : Security
>     {
>         public override double GetSecurityValue()
>         {
>             return YearsMaturity \* Interest;
>         }
>         public double Interest { get; set; }
>         public int YearsMaturity { get; set; }
>     }

**Comment by apryiomka** [comment]
And the DB context configuration:
` public class GuidantDbContext : DbContext
    {

```
    public GuidantDbContext() : base ("DefaultConnection") { }
    public DbSet<Portfolio> Portfolios { get; set; }
    public DbSet<Security> Securities { get; set; }
    public DbSet<Stock> Stocks { get; set; }
    public DbSet<Fund> Funds { get; set; }
    public DbSet<Bond> Bonds { get; set; }
    protected override void OnModelCreating(DbModelBuilder modelBuilder)
    {
        modelBuilder.Entity<Portfolio>().HasKey(p => p.Id);
        modelBuilder.Entity<Security>().HasKey(s => s.Id);
        modelBuilder.Entity<Fund>().ToTable("Funds");
        modelBuilder.Entity<Stock>().ToTable("Stocks");
        modelBuilder.Entity<Bond>().ToTable("Bonds");
        modelBuilder.Entity<Portfolio>().
            HasMany(c => c.Securities).
            WithMany(p => p.Protfolios).
            Map(
                m =>
                {
                    m.MapLeftKey("PortfolioId");
                    m.MapRightKey("SecurityId");
                    m.ToTable("PortfolioSecurities");
                });
    }
}
```

`

**Comment by vhrobles** [comment]
Makes sense! It's been a while since last time I've used EF so this is good to know. I've also read a little bit about fluent API
